### PR TITLE
#167377504 User should be able to vote on an article

### PR DESCRIPTION
**Pull request by Merkll**
#### What does this PR do?
This pull request implements the functionality for a user to vote on an article
#### Description of Task to be completed
* user upvote and downvote of an article
#### How should this be manually tested?
visit the storybook for the pull request

#### Any Background Context You want to provide?
N/A

#### What are the relevant pivotal tracker stories?

[#167377504](https://pivotaltracker.com/story/show/167377504)

### CODE REVIEW

File: src/assets/scss/style.scss [hunk 1]
```diff
@@ -1,5 +1,25 @@
 @import '~sass-rem';
+@import url('https://fonts.googleapis.com/css?family=Rubik:400,700&display=swap');
 
-.text {
-  color: #F00;
+$primarycolor: #FD8234;
```
Comment (iverenshaguy) [inline src/assets/scss/style.scss]: Please put these variables in a variables file that will be imported here. Thanks

File: src/components/Icon/icon.scss [hunk 2]
```diff
@@ -0,0 +1,50 @@
+$defaulticonstrokewidth: 0.03125rem;
```
Comment (iverenshaguy) [inline src/components/Icon/icon.scss]: The file should be `Icon.scss` not `icon.scss`

File: src/components/Icon/icon.scss [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+$activeiconfill: 'FD8234';
```
Comment (iverenshaguy) [inline src/components/Icon/icon.scss]: Kindly put these in variables

File: src/components/Icon/icon.scss [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+$defaulticonfill: 'C4C4C4';
+
+@function icon($iconName, $iconFill: $defaulticonfill) {
```
Comment (iverenshaguy) [inline src/components/Icon/icon.scss]: Please put this in a `functions.scss` file. Please also create a `mixins.scss` file for mixins

File: src/components/Icon/icon.spec.jsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import React from 'react';
```
Comment (iverenshaguy) [inline src/components/Icon/icon.spec.jsx]: Icon.spec.jsx

File: src/components/Icon/icon.stories.jsx [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// react libraries
```
Comment (iverenshaguy) [inline src/components/Icon/icon.stories.jsx]: Icon.stories.jsx

File: src/components/Icon/index.jsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import './icon.scss';
```
Comment (iverenshaguy) [inline src/components/Icon/index.jsx]:
```suggestion
import './Icon.scss';
```

File: src/components/Icon/index.jsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    * @name Icon propTypes
+    * @type {propTypes}
+    *
+    * @param {Object} props - React PropTypes
+    *
+    * @property {string} icon -icon type to render based on the css
+    * @property {Function} clickHandler -handles icon click and keypress event
+    * @property {Boolean} active -determines if icon is active
+    *
+    * @return {Array} React propTypes
```
Comment (iverenshaguy) [inline src/components/Icon/index.jsx]: Does this return an Array?

File: src/components/Icon/index.jsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    clickHandler: PropTypes.func.isRequired,
+    active: PropTypes.bool
+  }
+
+  static defaultProps = {
+    active: false
+  }
+
+  /**
+    * handles icon click event and calls the attach clicked handler
```
Comment (iverenshaguy) [inline src/components/Icon/index.jsx]:
```suggestion
    * Handles icon click event and calls the attach clicked handler
```

File: src/components/Icon/index.jsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    * @method
+    *
+    * @return {void}
+    */
+  onClickHandler = () => {
+    const { clickHandler, ...otherProps } = this.props;
+
+    clickHandler(otherProps);
```
Comment (iverenshaguy) [inline src/components/Icon/index.jsx]: Are we passing all other props to the clickHandler? What if this component needs to be extended

File: src/components/Icon/index.jsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  };
+
+  render() {
+    const { icon, active } = this.props;
+    return (
+      <button type="button" aria-label={icon} className={`icon icon-${icon} ${active && 'active'}`} onClick={this.onClickHandler} onKeyDown={this.onClickHandler} />
```
Comment (iverenshaguy) [inline src/components/Icon/index.jsx]: We should use the appropriate modifiers both here and in the SCSS file. Remember we are using BEM

```suggestion
      <button type="button" aria-label={icon} className={`icon icon--${icon} ${active && 'icon--active'}`} onClick={this.onClickHandler} onKeyDown={this.onClickHandler} />
```

File: src/components/Icon/index.jsx [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import React from 'react';
```
Comment (iverenshaguy) [inline src/components/Icon/index.jsx]: Please use the appropriate Import ordering and comments in all files

File: src/components/Vote/Vote.spec.jsx [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+  it('should use upVoteCount and downVoteCount props', () => {
+    const vote = shallow(<Provider store={store}><Vote articleId="900" upVoteCount={40} downVoteCount={40} activeVoteType="upVote" /></Provider>);
+    const { props } = vote.props().children;
+
+    expect(props.upVoteCount).toEqual(40);
+    expect(props.downVoteCount).toEqual(40);
+  });
+
+  it('check validity of state', () => {
```
Comment (iverenshaguy) [inline src/components/Vote/Vote.spec.jsx]:
```suggestion
  it('should check validity of state', () => {
```

File: src/components/Vote/Vote.stories.jsx [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+const withProvider = (story) => (
+  <Provider store={store}>
+    { story() }
+  </Provider>
+);
+
+storiesOf('Vote', module)
+  .addDecorator(withProvider)
+  .add('article vote', () => <Vote articleId="645a93d8-252a-4f5b-9a72-0a3af9323614" upVoteCount={30} activeVoteType="upVote" />);
```
Comment (iverenshaguy) [inline src/components/Vote/Vote.stories.jsx]: How does it look like when our activeVote is `downVote` or `nullVote`?

File: src/components/Vote/index.jsx [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+class Vote extends React.Component {
+  /**
+    * @name Vote propTypes
+    * @type {propTypes}
+    *
+    * @param {Object} props - React PropTypes
+    *
+    * @property {string} articleId -id of the article to vote on
+    * @property {string} activeVoteType - article voteType by the current user
+    * @property {Function} voteOnArticle -action creator to vote on an article
```
Comment (iverenshaguy) [inline src/components/Vote/index.jsx]:
```suggestion
    * @property {Function} voteOnArticle - action creator to vote on an article
```

File: src/components/Vote/index.jsx [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+ * @summary composes the upvote and the downvote icons along with the vote count
+ */
+class Vote extends React.Component {
+  /**
+    * @name Vote propTypes
+    * @type {propTypes}
+    *
+    * @param {Object} props - React PropTypes
+    *
+    * @property {string} articleId -id of the article to vote on
```
Comment (iverenshaguy) [inline src/components/Vote/index.jsx]:
```suggestion
    * @property {string} articleId - id of the article to vote on
```

File: src/components/Vote/index.jsx [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    * @property {string} activeVoteType - article voteType by the current user
+    * @property {Function} voteOnArticle -action creator to vote on an article
+    * @property {Function} articleVote -redux state map to props
```
Comment (iverenshaguy) [inline src/components/Vote/index.jsx]:
```suggestion
    * @property {Function} articleVote - redux state map to props
```

File: src/components/Vote/index.jsx [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    *
```
Comment (iverenshaguy) [inline src/components/Vote/index.jsx:38]: Please add the description for the other props

File: src/components/Vote/index.jsx [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    downVoteCount: 0
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = { activeVoteType: props.activeVoteType };
+  }
+
+  componentDidMount() {
+    const { initArticleVoteCount: articleVoteCount, upVoteCount, downVoteCount } = this.props;
```
Comment (iverenshaguy) [inline src/components/Vote/index.jsx]: Why name it something else and then rename in here?

File: src/components/Vote/index.jsx [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+
+    articleVoteCount({ upVoteCount, downVoteCount });
+  }
+
+  /**
+    * handles vote action
```
Comment (iverenshaguy) [inline src/components/Vote/index.jsx]:
```suggestion
    * Handles vote action
```

File: src/components/Vote/index.jsx [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+    * @method
+    *
+    * @param {Object} payload
+    * @param {string} payload.voteType - type of vote
+    *
+    * @return {void}
+    */
+  onVoteHandler = ({ voteType }) => {
+    const { voteOnArticle: voteOnArticleHandler, articleId } = this.props;
```
Comment (iverenshaguy) [inline src/components/Vote/index.jsx]: I'm not sure voteOnArticle needs to be renamed here

File: src/components/Vote/index.jsx [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+    const { activeVoteType } = this.state;
+
+    const vote = voteType === activeVoteType ? 'nullVote' : voteType;
+
+    voteOnArticleHandler({ articleId, voteType: vote, prevVote: activeVoteType });
+
+    this.setState((prevState) => ({ ...prevState, activeVoteType: vote }));
```
Comment (iverenshaguy) [inline src/components/Vote/index.jsx]: We don't need `...prevState`, React does this by default. Or is this a new API?
```suggestion
    this.setState({ activeVoteType: vote });
```

File: src/store/modules/vote/index.js [hunk 23]
```diff
@@ -0,0 +1,50 @@
+import { call, put, takeLatest } from 'redux-saga/effects';
+
+import { VOTE_ON_ARTICLE, VOTE_SUCCESSFULL, ARTICLE_VOTE_COUNT } from 'modules/vote/types';
```
Comment (iverenshaguy) [inline src/store/modules/vote/index.js]: Please let's change the action types to the below i.e.

VERB_NOUN_ACTION i.e. VOTE_ARTICLE_REQUEST, VOTE_ARTICLE_FAILURE, VOTE_ARTICLE_SUCCESS etc

```suggestion
import { VOTE_ARTICLE_REQUEST, VOTE_ARTICLE_SUCCESS, GET_VOTE_COUNT_REQUEST } from 'modules/vote/types';
```

Comment (iverenshaguy) [reply src/store/modules/vote/index.js]: Please correct throughout the file

File: src/store/modules/vote/vote.actions.test.js [hunk 24]
```diff
@@ -0,0 +1,19 @@
+import { voteOnArticle, onVoteSuccess } from './index';
+import { VOTE_ON_ARTICLE, VOTE_SUCCESSFULL } from './types';
+
+
+it('should return VOTE_ON_ARTICLE type', () => {
```
Comment (iverenshaguy) [inline src/store/modules/vote/vote.actions.test.js]: Wrap in a describe block. All tests file should have a root describe block

File: src/store/rootReducer.js [hunk 25]
```diff
@@ -2,8 +2,8 @@
 import { combineReducers } from 'redux';
 
 // modules
-import demo from 'modules/demo';
+import articleVote from 'modules/vote';
```
Comment (iverenshaguy) [inline src/store/rootReducer.js]: Maybe the module should also be called `articleVote` since there'll be `commentVote`

File: src/utils/axios.js [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+ * An axios proxy to attach the base url and authentication token
+ */
+import axios from 'axios';
+
+const axiosInstance = axios.create({
+  baseURL: process.env.API_ROUTE
+});
+
+axiosInstance.interceptors.request.use((config) => {
+  config.headers.Authorization = `BEARER ${localStorage.getItem('token')}`;
```
Comment (iverenshaguy) [inline src/utils/axios.js]: Could you also add a way for axios to automatically reject error responses?  You can intercept the response to do this

File: src/components/Vote/index.jsx [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+
+    return (
+      <div className="vote">
+        <div className="vote__wrapper">
+          <Icon icon="upvote" voteType="upVote" clickHandler={this.onVoteHandler} active={upVoteActive} />
+          <span className={`${upVoteActive && 'active'}`}>{upVoteCount}</span>
+        </div>
+        <div className="vote__wrapper">
+          <Icon icon="downvote" voteType="downVote" clickHandler={this.onVoteHandler} active={activeVoteType !== 'nullVote' && !upVoteActive} />
+          <span className={`${activeVoteType !== 'nullVote' && !upVoteActive && 'active'}`}>{downVoteCount}</span>
```
Comment (iverenshaguy) [inline src/components/Vote/index.jsx]: Let's create a class name formatter similar to the `classnames` package for these sort of use cases

File: src/utils/axios.js (hunk 26 above)
Comment (iverenshaguy) [reply src/utils/axios.js]: Please rename this to http i.e create a module for this i.e. http/index.js and also add the test from utils.spec.js to http/http.spec.js. Thanks

File: src/utils/utils.spec.js [hunk 28]
```diff
@@ -0,0 +1,21 @@
+import axios from 'utils/axios';
+
+describe('Axios HTTP client', () => {
+  it('should have a base url attached', () => {
+    expect(axios.defaults.baseURL).toBeDefined();
+  });
+
+  it('should attached authentication token', () => {
```
Comment (iverenshaguy) [inline src/utils/utils.spec.js]:
```suggestion
  it('should attach the authentication token', () => {
```

**Review by iverenshaguy — CHANGES_REQUESTED** [review]
Really great job. Please review my comments

**Comment by iverenshaguy** [comment]
@Bastieno Please why are these review apps not being deployed?

File: src/components/Vote/index.jsx (hunk 19 above)
Comment (Merkll) [reply src/components/Vote/index.jsx]: its used in the connect util to mapdispatchtoprop and eslint complains that its already declared in the upper scope

**Comment by iverenshaguy** [comment]
Please how do I test this feature?

File: src/utils/axios.js (hunk 26 above)
Comment (Merkll) [reply src/utils/axios.js]: should this still be in the utils folder?

Comment (Merkll) [reply src/utils/axios.js]: axios automatically throws an unsucessfull request as an error

File: src/components/Icon/index.jsx (hunk 10 above)
Comment (Merkll) [reply src/components/Icon/index.jsx]: used to pass back all props which can be used to track the click activities

File: src/utils/axios.js (hunk 26 above)
Comment (iverenshaguy) [reply src/utils/axios.js]: Really? So how come Gilbert's catch block isn't catching 400s and 500s?

**Comment by Merkll** [comment]
> Please how do I test this feature?

It isnt a standalone feature as an article page is needed

File: src/utils/axios.js (hunk 26 above)
Comment (Merkll) [reply src/utils/axios.js]: He wrapped it in a try/catch

**Comment by Merkll** [comment]
Requested feedback implemented.

File: src/utils/axios.js (hunk 26 above)
Comment (iverenshaguy) [reply src/utils/axios.js]: Isn't that why it should catch the errors?

Comment (Merkll) [reply src/utils/axios.js]: He wrapped the actual Axios request and the saga.

**Review by iverenshaguy — APPROVED** [review]

**Pull request merged**